typescript/infra/test/govern.hardhat-test.ts

```diff
@@ -0,0 +1,217 @@
+  HyperlaneAppGovernor,
+} from '../src/govern/HyperlaneAppGovernor';
+
+export class TestApp extends HyperlaneApp<{}> {}
+
+export class TestChecker extends HyperlaneAppChecker<TestApp, OwnableConfig> {
+  async checkChain(_: string): Promise<void> {
+    this.addViolation({
+      chain: Chains.test2,
+      type: 'test',
+      expected: 0,
+      actual: 1,
+    });
+  }
+}
+
+export class HyperlaneTestGovernor extends HyperlaneAppGovernor<
+  TestApp,
+  OwnableConfig
+> {
+  protected async mapViolationsToCalls() {
+    return;
+  }
+
+  mockPushCall(chain: string, call: AnnotatedCallData): void {
+    this.pushCall(chain, call);
+  }
+
+  async govern(_ = true, chain?: ChainName) {
+    // 2. For each call, infer how it should be submitted on-chain.
+    await this.inferCallSubmissionTypes();
+
+    // 3. Prompt the user to confirm that the count, description,
+    // and submission methods look correct before submitting.
+    const chains = chain ? [chain] : Object.keys(this.calls);
+    for (const chain of chains) {
+      await this.mockSendCalls(chain, this.calls[chain]);
+    }
+  }
+
+  protected async mockSendCalls(
+    chain: ChainName,
+    calls: CallData[],
+  ): Promise<void> {
+    for (const call of calls) {
+      await this.checker.multiProvider.sendTransaction(chain, {
+        to: call.to,
+        data: call.data,
+        value: call.value,
+      });
+    }
+  }
+}
+
+describe('ICA governance', async () => {
+  const localChain = Chains.test1;
+  const remoteChain = Chains.test2;
+
+  let signer: SignerWithAddress;
+  let multiProvider: MultiProvider;
+  let accountConfig: AccountConfig;
+  let coreApp: TestCoreApp;
+  // let local: InterchainAccountRouter;
+  let remote: InterchainAccountRouter;
+  let routerConfig: ChainMap<RouterConfig>;
+  let contracts: HyperlaneContractsMap<InterchainAccountFactories>;
+  let icaApp: InterchainAccount;
+  let recipient: TestRecipient;
+  let accountOwner: Address;
+  let governor: HyperlaneTestGovernor;
+
+  before(async () => {
+    [signer] = await ethers.getSigners();
+    multiProvider = MultiProvider.createTestMultiProvider({ signer });
+    const ismFactoryDeployer = new HyperlaneProxyFactoryDeployer(multiProvider);
+    const ismFactory = new HyperlaneIsmFactory(
+      await ismFactoryDeployer.deploy(multiProvider.mapKnownChains(() => ({}))),
+      multiProvider,
+    );
+
+    coreApp = await new TestCoreDeployer(multiProvider, ismFactory).deployApp();
+    routerConfig = coreApp.getRouterConfig(signer.address);
+  });
+
+  beforeEach(async () => {
+    contracts = await new InterchainAccountDeployer(multiProvider).deploy(
+      routerConfig,
+    );
+    // local = contracts[localChain].interchainAccountRouter;
+    remote = contracts[remoteChain].interchainAccountRouter;
+    icaApp = new InterchainAccount(contracts, multiProvider);
+
+    accountConfig = {
+      origin: Chains.test1,
+      owner: signer.address,
+      localRouter: remote.address,
```

Comment:
imo it is quite easy to mess up the address passed here
esp because of the "origin" and "local" language

Reply:
it's origin in the config, local in the deployers, governors.